DietPi-Arr_to_RAM | Add enable/disable a systemd unit to do copy tasks automatically

Hi @MichaIng this is not bug or feature request, just a question.

I want to make a service that will run a script that Fourdee made (this script will copy database files to ram, because this files don't let hdd to sleep)

https://github.com/MichaIng/DietPi/issues/2223#issuecomment-441082652

And I want that service to start before radarr and sonarr, what i need to change?

@MichaIng after i downloaded file
```
wget https://raw.githubusercontent.com/MichaIng/DietPi/arr_to_ram/dietpi/misc/dietpi-arr_to_RAM -O /DietPi/dietpi/misc/dietpi-arr_to_RAM
chmod +x /DietPi/dietpi/misc/dietpi-arr_to_RAM
```
i enabled service `/DietPi/dietpi/misc/dietpi-arr_to_RAM enable`
and after one minute i got this output
```
root@DietPi:~# /DietPi/dietpi/misc/dietpi-arr_to_RAM enable
Job for dietpi-arr_to_RAM.service failed because a timeout was exceeded.
See "systemctl status dietpi-arr_to_RAM.service" and "journalctl -xe" for details.
2019-04-09 00:26:56 | DietPi-Arr_to_RAM: [ERROR] An issue has occurred. Please check the log for details: /var/tmp/dietpi/logs/dietpi-arr_to_RAM.log
```
after second try i checked systemctl status
```root@DietPi:~# systemctl status dietpi-arr_to_RAM.service
● dietpi-arr_to_RAM.service - DietPi-Arr_to_RAM
   Loaded: loaded (/etc/systemd/system/dietpi-arr_to_RAM.service; enabled; vendor preset: enabled)
   Active: failed (Result: timeout) since Tue 2019-04-09 00:30:53 CEST; 1min 15s ago
  Process: 2379 ExecStart=/bin/dash -c /DietPi/dietpi/misc/dietpi-arr_to_RAM 1 2>&1 >> /var/tmp/dietpi/logs/dietpi-arr_to_RAM.log (code=killed, signal=TERM)

Apr 09 00:29:23 DietPi systemd[1]: Starting DietPi-Arr_to_RAM...
Apr 09 00:30:53 DietPi systemd[1]: dietpi-arr_to_RAM.service: Start operation timed out. Terminating.
Apr 09 00:30:53 DietPi systemd[1]: Failed to start DietPi-Arr_to_RAM.
Apr 09 00:30:53 DietPi systemd[1]: dietpi-arr_to_RAM.service: Unit entered failed state.
Apr 09 00:30:53 DietPi systemd[1]: dietpi-arr_to_RAM.service: Failed with result 'timeout'.
```
and var/log - it says already linked because i tried dietpi-arr_to_RAM enable two times
```
2019-04-09 00:29:23 | DietPi-Arr_to_RAM: Radarr detected
2019-04-09 00:29:23 | DietPi-Arr_to_RAM: Stopping Radarr service...
2019-04-09 00:29:23 | DietPi-Arr_to_RAM: Linking files to RAM...
2019-04-09 00:29:23 | DietPi-Arr_to_RAM: /mnt/dietpi_userdata/radarr/nzbdrone.db already linked to RAM (/tmp/radarr_db_link/nzbdrone.db). Skipping this file...
2019-04-09 00:29:23 | DietPi-Arr_to_RAM: /mnt/dietpi_userdata/radarr/nzbdrone.db-shm already linked to RAM (/tmp/radarr_db_link/nzbdrone.db-shm). Skipping this file...
2019-04-09 00:29:23 | DietPi-Arr_to_RAM: /mnt/dietpi_userdata/radarr/nzbdrone.db-wal already linked to RAM (/tmp/radarr_db_link/nzbdrone.db-wal). Skipping this file...
```

after that i hit `/DietPi/dietpi/misc/dietpi-arr_to_RAM 0 # Restore to disk` and restared system and everything loaded fine
```
root@DietPi:~# systemctl status dietpi-arr_to_RAM.service
● dietpi-arr_to_RAM.service - DietPi-Arr_to_RAM
   Loaded: loaded (/etc/systemd/system/dietpi-arr_to_RAM.service; enabled; vendor preset: enabled)
   Active: active (exited) since Tue 2019-04-09 00:35:25 CEST; 3min 2s ago
  Process: 309 ExecStart=/bin/dash -c /DietPi/dietpi/misc/dietpi-arr_to_RAM 1 2>&1 >> /var/tmp/dietpi/logs/dietpi-arr_to_RAM.log (code=exited, status=0/SUCCESS)
   CGroup: /system.slice/dietpi-arr_to_RAM.service

Apr 09 00:35:24 DietPi systemd[1]: Starting DietPi-Arr_to_RAM...
Apr 09 00:35:25 DietPi systemd[1]: Started DietPi-Arr_to_RAM.
```
```
2019-04-09 00:35:24 | DietPi-Arr_to_RAM: Radarr detected
2019-04-09 00:35:24 | DietPi-Arr_to_RAM: Linking files to RAM...
2019-04-09 00:35:24 | DietPi-Arr_to_RAM: Linking /mnt/dietpi_userdata/radarr/nzbdrone.db to RAM (/tmp/radarr_db_link/nzbdrone.db)...
2019-04-09 00:35:24 | DietPi-Arr_to_RAM: Linking /mnt/dietpi_userdata/radarr/nzbdrone.db-shm to RAM (/tmp/radarr_db_link/nzbdrone.db-shm)...
2019-04-09 00:35:24 | DietPi-Arr_to_RAM: Linking /mnt/dietpi_userdata/radarr/nzbdrone.db-wal to RAM (/tmp/radarr_db_link/nzbdrone.db-wal)...
2019-04-09 00:35:24 | DietPi-Arr_to_RAM: Sonarr detected
2019-04-09 00:35:24 | DietPi-Arr_to_RAM: Linking files to RAM...
2019-04-09 00:35:24 | DietPi-Arr_to_RAM: Linking /mnt/dietpi_userdata/sonarr/nzbdrone.db to RAM (/tmp/sonarr_db_link/nzbdrone.db)...
2019-04-09 00:35:25 | DietPi-Arr_to_RAM: Linking /mnt/dietpi_userdata/sonarr/nzbdrone.db-shm to RAM (/tmp/sonarr_db_link/nzbdrone.db-shm)...
2019-04-09 00:35:25 | DietPi-Arr_to_RAM: Linking /mnt/dietpi_userdata/sonarr/nzbdrone.db-wal to RAM (/tmp/sonarr_db_link/nzbdrone.db-wal)...

```

@MichaIng no, files were on disk.. i brought them back with `dietpi-arr_to_RAM 0` before downloading new dietpi-arr_to_RAM file.. i deleted `rm /etc/systemd/system/dietpi-arr_to_RAM.service` few days ago when service didn't work normaly, and i see that the new service have the same name dietpi-arr_to_RAM.service, maybe that was the problem..i restarted few times dietpi between these two actions..

I saw logs after i enabled service and before restarted dietpi, and i saw that only radarr files were moved to ram..maybe problem was with sonarr service didn't stop.. 
```
2019-04-09 00:29:23 | DietPi-Arr_to_RAM: Radarr detected
2019-04-09 00:29:23 | DietPi-Arr_to_RAM: Stopping Radarr service...
2019-04-09 00:29:23 | DietPi-Arr_to_RAM: Linking files to RAM...
2019-04-09 00:29:23 | DietPi-Arr_to_RAM: /mnt/dietpi_userdata/radarr/nzbdrone.db already linked to RAM (/tmp/radarr_db_link/nzbdrone.db). Skipping this file...
2019-04-09 00:29:23 | DietPi-Arr_to_RAM: /mnt/dietpi_userdata/radarr/nzbdrone.db-shm already linked to RAM (/tmp/radarr_db_link/nzbdrone.db-shm). Skipping this file...
2019-04-09 00:29:23 | DietPi-Arr_to_RAM: /mnt/dietpi_userdata/radarr/nzbdrone.db-wal already linked to RAM (/tmp/radarr_db_link/nzbdrone.db-wal). Skipping this file...
```
There is no mention of sonarr, i tried two times and saw logs, and both times logs was only displaying radarr actions.. i will try later today to disable service, and enable it again to see what will happen..

Edit: i quess it has something to do with sonarr
I disabled `/DietPi/dietpi/misc/dietpi-arr_to_RAM disable` everything went ok
```
2019-04-09 14:52:40 | DietPi-Arr_to_RAM: Radarr detected
2019-04-09 14:52:40 | DietPi-Arr_to_RAM: Stopping Radarr service...
2019-04-09 14:52:40 | DietPi-Arr_to_RAM: Restoring files from RAM (/tmp/radarr_db_link) to disk (/mnt/dietpi_userdata/radarr)...
2019-04-09 14:52:44 | DietPi-Arr_to_RAM: Sonarr detected
2019-04-09 14:52:44 | DietPi-Arr_to_RAM: Stopping Sonarr service...
2019-04-09 14:52:44 | DietPi-Arr_to_RAM: Restoring files from RAM (/tmp/sonarr_db_link) to disk (/mnt/dietpi_userdata/sonarr)...
```
and after restart i enabled it again, and again same error
```
root@DietPi:~# /DietPi/dietpi/misc/dietpi-arr_to_RAM enable
Created symlink /etc/systemd/system/multi-user.target.wants/dietpi-arr_to_RAM.service → /etc/systemd/system/dietpi-arr_to_RAM.service.
Job for dietpi-arr_to_RAM.service failed because a timeout was exceeded.
See "systemctl status dietpi-arr_to_RAM.service" and "journalctl -xe" for details.
2019-04-09 14:58:21 | DietPi-Arr_to_RAM: [ERROR] An issue has occurred. Please check the log for details: /var/tmp/dietpi/logs/dietpi-arr_to_RAM.log
root@DietPi:~# nano /var/tmp/dietpi/logs/dietpi-arr_to_RAM.log
```
and in log file there is nothing about sonarr
```
2019-04-09 14:56:51 | DietPi-Arr_to_RAM: Radarr detected
2019-04-09 14:56:51 | DietPi-Arr_to_RAM: Stopping Radarr service...
2019-04-09 14:56:51 | DietPi-Arr_to_RAM: Linking files to RAM...
2019-04-09 14:56:51 | DietPi-Arr_to_RAM: Linking /mnt/dietpi_userdata/radarr/nzbdrone.db to RAM (/tmp/radarr_db_link/nzbdrone.db)...
2019-04-09 14:56:51 | DietPi-Arr_to_RAM: Linking /mnt/dietpi_userdata/radarr/nzbdrone.db-shm to RAM (/tmp/radarr_db_link/nzbdrone.db-shm)...
2019-04-09 14:56:51 | DietPi-Arr_to_RAM: Linking /mnt/dietpi_userdata/radarr/nzbdrone.db-wal to RAM (/tmp/radarr_db_link/nzbdrone.db-wal)...
```
and again after bringing radarr files from ram and restart, everything went ok

@Dr0bac 
Jep just did some final polishing and passed my tests successfully.

I added the possibility to update the backups via:
```
/DietPi/dietpi/misc/dietpi-arr_to_RAM 2
```

Since unix permissions are required to do so, and Sonarr/Radarr/Lidarr run as their own limited user, it was required to allow this per-program. As this is generally a nice feature and to have everything more consistent, I added the possibility to do all tasks per-program optionally, e.g.:
```
/DietPi/dietpi/misc/dietpi-arr_to_RAM 2 radarr
```
will only update the Radarr databases. The same can be done to only link certain program databases to RAM or back to disk. The boot service still applies to all programs. This would have been too complicated for now an generally you either want that for all or non of the programs. Linking a single program to RAM or back most likely is only interesting for debug reasons or to fix errors, restore individual backups and such anyway.

Since systemd service start/stop requires root permissions, only the database backup updates can be done as sonarr/radarr/lidarr program user, and of course only for the single program. So what you can/need to do now is using the "Connect" > "Custom script" feature and add:
Command: `/DietPi/dietpi/misc/dietpi-arr_to_RAM`
Agrument: `2 radarr`

Just tested as well. Hitting "Test" does not really execute the script, not sure if it just checks existence and excitability? 🤔 But when adding any test resource with enable "On grab" trigger, then it indeed updates the database backup automatically. So should work with finished downloads/renames/updates well.

Btw: I wanted to implement automated adding the custom script trigger via the script itself. But this information is within the database files and those cannot be manipulated from outside. There is an API available to perform certain tasks, but at least currently this does not include changing these settings. So this needs to be added manually. However not too hard through the GUI, I will add the info to our online docs, once released.
____________
Running `/DietPi/dietpi/misc/dietpi-arr_to_RAM` without argument or wrong (combination) will show you nice usage info now which should explain all features as well. Hope this is understandable, otherwise I am of course opened to polish/clarify some wording, if required:
```
root@VM-Stretch:/mnt/dietpi_userdata/radarr# /DietPi/dietpi/misc/dietpi-arr_to_RAM 2 radarrwhoopsie
2019-04-13 19:16:28 | DietPi-Arr_to_RAM: [ERROR] Invalid input program (radarrwhoopsie). Aborting...

Usage: /DietPi/dietpi/misc/dietpi-arr_to_RAM <command> [<program>]
Available commands:
  1 [<program>]         Link (program) database(s) to RAM
  2 [<program>]         Update (program) database backup(s)
  0 [<program>]         Store (program) database(s) back to disk
  enable                Enable Link to RAM on boot
  disable               Disable Link to RAM on boot
Supported programs:
  <empty>               Apply to all supported and installed programs
  sonarr                Apply to Sonarr database only
  radarr                Apply to Radarr database only
  lidarr                Apply to Lidarr database only
```

--- a/dietpi/misc/dietpi-arr_to_RAM
+++ b/dietpi/misc/dietpi-arr_to_RAM
@@ -143,7 +143,7 @@ $USAGE"; exit 1; }
 			if [[ ! -f $FP_DISK/dietpi-arr_to_RAM.sh ]]
 			then
 				Print "Creating $FP_DISK/dietpi-arr_to_RAM.sh to be used as ${i^} custom script"
-				echo -e '#!/bin/dash\n/boot/dietpi/misc/dietpi-arr_to_RAN 2 '"$i" > "$FP_DISK/dietpi-arr_to_RAM.sh" || { error=$? Print "Creating $FP_DISK/dietpi-arr_to_RAM.sh failed."; }
+				echo -e '#!/bin/dash\n/boot/dietpi/misc/dietpi-arr_to_RAM 2 '"$i" > "$FP_DISK/dietpi-arr_to_RAM.sh" || { error=$? Print "Creating $FP_DISK/dietpi-arr_to_RAM.sh failed."; }
 				chmod +x "$FP_DISK/dietpi-arr_to_RAM.sh" || { error=$? Print "Applying $FP_DISK/dietpi-arr_to_RAM.sh execute permissions failed."; }
 				chown "$i:dietpi" "$FP_DISK/dietpi-arr_to_RAM.sh" || { error=$? Print "Applying $FP_DISK/dietpi-arr_to_RAM.sh ownership failed."; }
 			fi

